protoc-gen-grpc-gateway/gengateway/template.go

```diff
@@ -207,6 +207,7 @@ var (
 	var protoReq {{.Method.RequestType.GoType .Method.Service.File.GoPkg.Path}}
 	var metadata runtime.ServerMetadata
 {{if .Body}}
+	{{.Body.AssignableExprPrep "protoReq"}}
 	if err := marshaler.NewDecoder(req.Body).Decode(&{{.Body.AssignableExpr "protoReq"}}); err != nil && err != io.EOF  {
```

Comment:
This is what generated bad code. The `AssignableExpr` would have emitted the prep statements, which would get incorrectly inlined in the call to `Decode`. By splitting out the prep statements, it now generates working code.